presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java

```diff
@@ -0,0 +1,205 @@
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.optimizations.joins.JoinGraph;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class EliminateCrossJoins
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        if (!SystemSessionProperties.isJoinReorderingEnabled(session)) {
+            return plan;
+        }
+
+        List<JoinGraph> joinGraphs = JoinGraph.buildFrom(plan);
+
+        for (int i = 0; i < joinGraphs.size(); i++) {
+            JoinGraph graph = joinGraphs.get(i);
+            List<Integer> joinOrder = getJoinOrder(graph);
+            if (isOriginalOrder(joinOrder)) {
+                continue;
+            }
+
+            plan = rewriteWith(new Rewriter(idAllocator, graph, joinOrder), plan);
+            joinGraphs = JoinGraph.buildFrom(plan);
+        }
+
+        return plan;
+    }
+
+    public static boolean isOriginalOrder(List<Integer> joinOrder)
+    {
+        for (int i = 0; i < joinOrder.size(); i++) {
+            if (joinOrder.get(i) != i) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Given JoinGraph determine the order of joins between graph nodes
+     * by traversing JoinGraph. Any graph traversal algorithm could be used
+     * here (like BFS or DFS), but we use PriorityQueue to preserve
+     * original JoinOrder as mush as it is possible. PriorityQueue returns
+     * next nodes to join in order of their occurrence in original Plan.
+     */
+    public static List<Integer> getJoinOrder(JoinGraph graph)
+    {
+        ImmutableList.Builder<PlanNode> joinOrder = ImmutableList.builder();
+
+        Map<PlanNode, Integer> priorities = new HashMap<>();
```

Comment:
Same comment re: equals/hashCode for PlanNode. You may want to use PlanNodeId as the key, instead.